lib/spack/spack/main.py

```diff
@@ -99,16 +101,21 @@ def set_working_dir():
         spack_working_dir = spack.paths.prefix
 
 
+def all_commands():
+    """Return all top level commands available with Spack."""
+    return spack.cmd.all_commands(_all_commands)
+
+
 def add_all_commands(parser):
     """Add all spack subcommands to the parser."""
-    for cmd in spack.cmd.all_commands():
+    for cmd in spack.cmd.all_commands(_all_commands):
```

Comment:
As per the suggestion above I think you could have called the other function something else and avoided moving `all_commands` to this module.

Reply:
The presence of the `_all_commands` cache variable and its relevance specifically to top-level Spack commands rather than any sub-commands, is what motivated me to move the older function to `spack.main` as much as the naming clash. I can certainly rename the more general function (suggestions welcome), but it seems to me that notwithstanding that, there remains a good reason to have the specific `all_commands()` function implementing the older behavior in `spack.main` rather than its original location of `spack.cmd`. Thoughts?